package com.yazikochesalna.fileservice.advice;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.multipart.support.MissingServletRequestPartException;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@RestControllerAdvice
public class GlobalExceptionHandler {

    private static final String TIMESTAMP = "timestamp";
    private static final String PATH = "path";
    private static final String URI_PREFIX = "uri=";
    private static final String DEFAULT_EXCEPTION_MESSAGE = "Произошла ошибка";

    @ExceptionHandler(Exception.class)
    public ResponseEntity<CustomErrorResponse> handleException(Exception ex, WebRequest request) {
        Map<String, Object> details = new HashMap<>();
        details.put(TIMESTAMP, LocalDateTime.now());
        details.put(PATH, request.getDescription(false).replace(URI_PREFIX, ""));

        String message = ex.getMessage() != null ? ex.getMessage() : DEFAULT_EXCEPTION_MESSAGE;

        return new ResponseEntity<>(
                new CustomErrorResponse(message, details),
                getStatus(ex)
        );
    }

    private HttpStatus getStatus(Exception ex) {
        ResponseStatus responseStatus = ex.getClass().getAnnotation(ResponseStatus.class);
        if (responseStatus != null) {
            return responseStatus.value();
        }

        return HttpStatus.INTERNAL_SERVER_ERROR;
    }

    @ExceptionHandler(MissingServletRequestPartException.class)
    public ResponseEntity<CustomErrorResponse> handleMissingPart(MissingServletRequestPartException ex, WebRequest request) {
        Map<String, Object> details = new HashMap<>();
        details.put(TIMESTAMP, LocalDateTime.now());
        details.put(PATH, request.getDescription(false).replace(URI_PREFIX, ""));

        return new ResponseEntity<>(
                new CustomErrorResponse(ex.getMessage(), details),
                HttpStatus.BAD_REQUEST
        );
    }


}